### DISCUSSION TITLE
ant5中，menu组件动态渲染icon问题

### DISCUSSION THREAD

**Discussion by YuriPLSk** [ant-design/ant-design#41981] [General]
**react17 + antd5开发**

因为目前antd5移除了menu.item子级，而是通过menu组件的items属性来控制整个菜单的内容
eg:<Menu items={{ [菜单数组]  }}>
我的痛点是在渲染菜单前我需要对items处理，而icon是作为一个字符串存在数据库中，所以数据返回后我需要改为reactNode类型
经排查后

>  import * as Icon from "@ant-design/icons" //引入*以满足动态渲染
>  ... ...
>  list.map(item=>{
>     item.icon = React.createElement(Icons[item.icon])
>  })

但是前端却报错：

>  element type is invalid : expect a string (...) or a class/fuction (for...) but got undefined ...

前一天下班认为是自己的方法不对困扰许久，结果今早上再次看到报错就认真读了一遍，这个地方需要的是一个函数，类或者字符串，而我写的是一个reactNode，所以我直接就是一个

>  item.icon = **()=>**React.createElement(Icons[item.icon])

结果，成了....
因为自己粗心钻到死胡同，血泪经验供大家参考，

**Comment by Wxh16144** [comment]
能分享踩坑历程可太棒了 👍 

不过我有几个困惑的点:

1. `list.map` 在迭代的时候不应该是通过 `item.xxx` 去改变内容, 而是通过 `map` 的 `callback return` 一个新的 `item`
2. `React.createElement` 创建的是一个 ReactNode, 和 `()=> React.createElement(Icons[item.icon])` 感觉应该是同理的吧 🤔?

最终代码这样貌似也能跑

```tsx
list.map(item=>({
  ...item,
  icon: React.createElement(Icons[item.icon])
}));
```

**Reply by Wxh16144** [reply]
难道是 React 17 不同的差异?

**Reply by YuriPLSk** [reply]
谢谢回答
因为我在内网开发，所以没办法贴图以更清晰展示我的代码
1.我用了一个变量来接收`list.map()`的返回值，但是的确若要对数组item本身做操作使用`list.forEach()`更为合理和快捷
2.我使用` icon: React.createElement(Icons[item.icon])`，在编译阶段就会抛出上文的异常；使用函数返回后发现编译通过，但console报错`fuctions are not valid as a React Child . This may happen if you return a Component instead of <Component /> from render ... `

我可能跳入了另一个坑 sad

**Comment by YuriPLSk** [comment]
问题找到了，我的list mock结构大致如下
`[{
id:"XXX",
icon:"/users" ||  "<users>" 
},
{...}
]`
而后我把数据库存的icon字段改为`users`，
原代码使用`icon: React.createElement(Icons[item.icon])`即可
